pkg/sfu/sendsidebwe/packet_tracker.go

```diff
@@ -4,49 +4,94 @@ import (
 	"sync"
 	"time"
 
+	"github.com/MicahParks/peakdetect"
+	"github.com/frostbyte73/core"
+	"github.com/gammazero/deque"
 	"github.com/livekit/protocol/logger"
+	"github.com/pion/rtcp"
 )
 
 type packetInfo struct {
-	sendTime    time.Time
+	sendTime time.Time
+	// SSBWE-TODO: may need a feedback report time to detect stale reports
 	arrivalTime int64
 	headerSize  uint16
 	payloadSize uint16
+	// SSBWE-TODO: possibly add the following fields - pertaining to this packet,
+	// idea is to be able to traverse back and find last packet with clean signal(s)
+	// in order to figure out bitrate at which congestion triggered.
+	//    MPETau - Mean Percentage Error trend - potentially an early signal of impending congestion
+	//    PeakDetectorSignal - A sustained peak could indicate definite congestion
+	//    AcknowledgedBitrate - Bitrate acknowledged by feedback
+	//    ProbePacketInfo - When probing, these packets could be analyzed separately to check if probing is successful
+}
+
+type feedbackInfo struct {
+	baseSN     uint16
+	numPackets uint16
 }
 
 type PacketTracker struct {
 	logger logger.Logger
 
-	lock        sync.RWMutex
-	initialized bool
-	highestSN   uint16
-	packetInfos [1 << 16]packetInfo
+	lock            sync.RWMutex
+	sentInitialized bool
+	highestSentSN   uint16
+
+	ackedInitialized bool
+	highestAckedSN   uint16
+
+	// SSBWE-TODO: make this a ring buffer as a lot more fields are needed
+	packetInfos   [1 << 16]packetInfo
```

Comment:
Still a large array here. But, as more things get stored in packet info, this should become a much smaller slice (maybe like 2K - 4K packets) or some other better implementation (maybe a deque or something like https://github.com/GavinClarke0/lockless-generic-ring-buffer